Error with else and if statement

Hi!

Snakefmt failes while parsing my workflow:

```
 if config["programs"]["Duplicates"] == "fgbio":
    include: "../rules/Alignment/fgbio.smk"
else:
    if config["programs"]["markduplicate"] == "GPU":
        include: "../rules/Alignment/GPU_alignment.smk"
    else:
        include: "../rules/Alignment/bwa-mem.smk"
        include: "../rules/Alignment/MarkDuplicates.smk"
```

with the following message

> [ERROR] In file "src/Snakemake/workflow/Twist_DNA_workflow.smk":  'InvalidPython: Got error:
> ```
> Cannot parse: 1:2: if:
> ```
> while formatting code with black.'
> [INFO] In file "src/Snakemake/workflow/Twist_DNA_workflow.smk":  1 file(s) raised parsing errors

running black manually gives me no errors. 

if I remove the if statement after else I get no errors.

```
if config["programs"]["Duplicates"] == "fgbio":
    include: "../rules/Alignment/fgbio.smk"
else:
    include: "../rules/Alignment/GPU_alignment.smk"
```

--- a/snakefmt/formatter.py
+++ b/snakefmt/formatter.py
@@ -28,7 +28,7 @@ triple_quote_matcher = re.compile(
     r"^\s*(\w?\"{3}.*?\"{3})|^\s*(\w?'{3}.*?'{3})", re.DOTALL | re.MULTILINE
 )
 contextual_matcher = re.compile(
-    r"(.*)^(if|elif|else|with|for|while)(.*)(:.*)", re.S | re.M
+    r"(.*)^(if|elif|else|with|for|while)([^:]*)(:.*)", re.S | re.M
 )
 
 
@@ -88,9 +88,7 @@ class Formatter(Parser):
                 else:
                     test_substitute = f"{used_keyword} a"
                 to_format = (
-                    f"{re_match.group(1)}{test_substitute}"
-                    f"{re_match.group(4)}"
-                    f"{TAB * self.context.cur_indent}pass"
+                    f"{re_match.group(1)}{test_substitute}" f"{re_match.group(4)}pass"
                 )
                 formatted = self.run_black_format_str(to_format, self.target_indent)
                 re_rematch = contextual_matcher.match(formatted)
@@ -101,10 +99,14 @@ class Formatter(Parser):
                 )
                 formatted_lines = formatted.splitlines(keepends=True)
                 formatted = "".join(formatted_lines[:-1])  # Remove the 'pass' line
+                indent_lines = sum(
+                    [line.strip().endswith(":") for line in formatted_lines]
+                )
+                code_indent = max(self.context.cur_indent - indent_lines, 0)
             else:
                 formatted = self.run_black_format_str(self.buffer, self.target_indent)
-            code_indent = max(self.context.cur_indent - 1, 0)
-            formatted = f"{TAB * code_indent}{formatted}"
+                code_indent = max(self.context.cur_indent - 1, 0)
+            formatted = textwrap.indent(formatted, f"{TAB * code_indent}")
 
         # Re-add newline removed by black for proper parsing of comments
         if self.buffer.endswith("\n\n"):

--- a/snakefmt/formatter.py
+++ b/snakefmt/formatter.py
@@ -10,7 +10,7 @@ from snakefmt.config import PathLike, read_black_config
 from snakefmt.exceptions import InvalidParameterSyntax, InvalidPython
 from snakefmt.logging import Warnings
 from snakefmt.parser.grammar import SnakeRule
-from snakefmt.parser.parser import Parser
+from snakefmt.parser.parser import Parser, comment_start
 from snakefmt.parser.syntax import (
     COMMENT_SPACING,
     TAB,
@@ -30,10 +30,6 @@ triple_quote_matcher = re.compile(
 contextual_matcher = re.compile(
     r"(.*)^(if|elif|else|with|for|while)([^:]*)(:.*)", re.S | re.M
 )
-
-
-def comment_start(string: str) -> bool:
-    return string.lstrip().startswith("#")
 
 
 class Formatter(Parser):

--- a/snakefmt/formatter.py
+++ b/snakefmt/formatter.py
@@ -46,7 +46,6 @@ class Formatter(Parser):
         self.result: str = ""
         self.lagging_comments: str = ""
         self.no_formatting_yet: bool = True
-        self.last_recognised_keyword = ""
 
         self.black_mode = read_black_config(black_config_file)
 
@@ -99,13 +98,9 @@ class Formatter(Parser):
                 )
                 formatted_lines = formatted.splitlines(keepends=True)
                 formatted = "".join(formatted_lines[:-1])  # Remove the 'pass' line
-                indent_lines = sum(
-                    [line.strip().endswith(":") for line in formatted_lines]
-                )
-                code_indent = max(self.context.cur_indent - indent_lines, 0)
             else:
                 formatted = self.run_black_format_str(self.buffer, self.target_indent)
-                code_indent = max(self.context.cur_indent - 1, 0)
+            code_indent = self.context.code_indent
             formatted = textwrap.indent(formatted, f"{TAB * code_indent}")
 
         # Re-add newline removed by black for proper parsing of comments
@@ -113,7 +108,8 @@ class Formatter(Parser):
             if comment_start(self.buffer.rstrip().splitlines()[-1]):
                 formatted += "\n"
         # Only stick together separated single-parm keywords when separated by comments
-        if not all(map(comment_start, self.buffer.splitlines())):
+        buffer_is_all_comments = all(map(comment_start, self.buffer.splitlines()))
+        if not buffer_is_all_comments:
             self.last_recognised_keyword = ""
         self.add_newlines(self.target_indent, formatted, final_flush, in_global_context)
         self.buffer = ""
@@ -284,8 +280,7 @@ class Formatter(Parser):
         have_only_comment_lines = comment_break == 0
         if not have_only_comment_lines or final_flush:
             collate_same_singleparamkeyword = (
-                cur_indent == 0
-                and context is not None
+                context is not None
                 and context.keyword_name == self.last_recognised_keyword
                 and issubclass(context.__class__, SingleParam)
             )

--- a/snakefmt/parser/parser.py
+++ b/snakefmt/parser/parser.py
@@ -41,6 +41,7 @@ class Parser(ABC):
         self.context_stack = [self.grammar]
         self.snakefile: TokenIterator = snakefile
         self.from_python: bool = False
+        self.last_recognised_keyword: str = ""
 
         status = self.context.get_next_queriable(self.snakefile)
         self.buffer = status.buffer
@@ -52,7 +53,6 @@ class Parser(ABC):
             if status.eof:
                 break
 
-            self.from_python = False
             keyword = status.token.string
             if self.vocab.recognises(keyword):
                 if status.indent > self.target_indent:
@@ -60,10 +60,15 @@ class Parser(ABC):
                         self.from_python = True
                     else:  # Over-indented context gets reset
                         self.context.cur_indent = max(self.target_indent - 1, 0)
+                elif self.from_python:
+                    self.from_python = False
+                    # We are exiting python context, so force spacing out keywords
+                    self.last_recognised_keyword = ""
                 self.flush_buffer(
                     from_python=self.from_python,
                     in_global_context=self.in_global_context,
                 )
+                self.context.code_indent = None
                 status = self.process_keyword(status, self.from_python)
             else:
                 if not self.context.accepts_python_code and not keyword[0] == "#":
@@ -73,6 +78,8 @@ class Parser(ABC):
                     )
                 else:
                     self.buffer += keyword
+                    if self.context.code_indent is None:
+                        self.context.code_indent = status.indent
                     status = self.context.get_next_queriable(self.snakefile)
                     self.buffer += status.buffer
             self.context.cur_indent = status.indent

--- a/snakefmt/parser/parser.py
+++ b/snakefmt/parser/parser.py
@@ -33,6 +33,10 @@ class Snakefile:
         self._buffered_tokens.append(token)
 
 
+def comment_start(string: str) -> bool:
+    return string.lstrip().startswith("#")
+
+
 class Parser(ABC):
     def __init__(self, snakefile: TokenIterator):
         self.grammar = Grammar(
@@ -71,14 +75,14 @@ class Parser(ABC):
                 self.context.code_indent = None
                 status = self.process_keyword(status, self.from_python)
             else:
-                if not self.context.accepts_python_code and not keyword[0] == "#":
+                if not self.context.accepts_python_code and not comment_start(keyword):
                     raise SyntaxError(
                         f"L{status.token.start[0]}: Unrecognised keyword '{keyword}' "
                         f"in {self.context.keyword_name} definition"
                     )
                 else:
                     self.buffer += keyword
-                    if self.context.code_indent is None:
+                    if self.context.code_indent is None and not comment_start(keyword):
                         self.context.code_indent = status.indent
                     status = self.context.get_next_queriable(self.snakefile)
                     self.buffer += status.buffer

--- a/snakefmt/parser/syntax.py
+++ b/snakefmt/parser/syntax.py
@@ -117,6 +117,7 @@ class Syntax:
         assert target_indent >= 0
         self.target_indent = target_indent
         self.cur_indent = max(self.target_indent - 1, 0)
+        self.code_indent = None
         self.comment = ""
         self.token = None
 

--- a/tests/test_formatter.py
+++ b/tests/test_formatter.py
@@ -380,16 +380,30 @@ class TestComplexPythonFormatting:
         formatter = setup_formatter(snakecode)
         assert formatter.get_formatted() == snakecode
 
-    def test_nested_ifelse_statements_indentation_maintained(self):
+    def test_nested_ifelse_statements(self):
         snakecode = (
             'if config["a"] is None:\n\n'
             f'{TAB * 1}include: "module_a_none.smk"\n\n\n'
+            f"else:\n"
             f'{TAB * 1}if config["b"] is None:\n\n'
             f'{TAB * 2}include: "module_b.smk"\n\n\n'
             f"{TAB * 1}else:\n\n"
-            f'{TAB * 2}include: "module_c.smk"\n\n\n'
-            "else:\n\n"
-            f'{TAB * 1}include: "module_a.smk"\n'
+            f'{TAB * 2}include: "module_c.smk"\n'
+        )
+        formatter = setup_formatter(snakecode)
+        assert formatter.get_formatted() == snakecode
+
+    def test_nested_ifelse_statements_multiple_python_lines(self):
+        snakecode = (
+            'if config["a"] is None:\n'
+            f"{TAB * 1}a = 1\n\n"
+            f'{TAB * 1}include: "module_a_none.smk"\n\n\n'
+            f"else:\n"
+            f'{TAB * 1}if config["b"] is None:\n\n'
+            f'{TAB * 2}include: "module_b.smk"\n\n\n'
+            f"{TAB * 1}else:\n"
+            f"{TAB * 2}b = 0\n\n"
+            f'{TAB * 2}include: "module_c.smk"\n'
         )
         formatter = setup_formatter(snakecode)
         assert formatter.get_formatted() == snakecode

--- a/tests/test_formatter.py
+++ b/tests/test_formatter.py
@@ -369,13 +369,22 @@ class TestComplexPythonFormatting:
         )
         assert formatter.get_formatted() == expected
 
-    def test_parameter_keywords_inside_python_code(self):
+    def test_indented_consecutive_snakemake_directives(self):
         snakecode = (
-            "if condition:\n\n"
-            f'{TAB * 1}include: "a"\n\n\n'
+            'if config["load"]:\n\n'
+            f'{TAB * 1}include: "module_a.smk"\n'
+            f'{TAB * 1}include: "module_b.smk"\n'
+        )
+        formatter = setup_formatter(snakecode)
+        assert formatter.get_formatted() == snakecode
+
+    def test_spaced_out_consecutive_dedented_directive(self):
+        snakecode = (
+            'if config["load"]:\n\n'
+            f'{TAB * 1}include: "module_a.smk"\n\n\n'
             f"else:\n\n"
-            f'{TAB * 1}include: "b"\n'
-            f'include: "c"\n'
+            f'{TAB * 1}include: "module_b.smk"\n\n\n'
+            f'include: "other.smk"\n'
         )
         formatter = setup_formatter(snakecode)
         assert formatter.get_formatted() == snakecode

--- a/tests/test_formatter.py
+++ b/tests/test_formatter.py
@@ -389,6 +389,17 @@ class TestComplexPythonFormatting:
         formatter = setup_formatter(snakecode)
         assert formatter.get_formatted() == snakecode
 
+    def test_comment_support_after_python_code(self):
+        snakecode = (
+            'if config["a"]:\n\n'
+            f'{TAB * 1}include: "module_a.smk"\n\n\n'
+            f'# include: "module_b.smk"\n\n\n'
+            f'if config["c"]:\n\n'
+            f'{TAB * 1}include: "module_c.smk"\n'
+        )
+        formatter = setup_formatter(snakecode)
+        assert formatter.get_formatted() == snakecode
+
     def test_nested_ifelse_statements(self):
         snakecode = (
             'if config["a"] is None:\n\n'

